File: java/cours/src/session2/td/heritageMultiple/IntPositif.java
package session2.td.heritageMultiple;

import session1.td.Nat;

public abstract class IntPositif implements Nat {
	
	private int val;

	public IntPositif(int val){
		if(val < 0)
			throw new IllegalArgumentException("Pas de Nat à patir d'un int négatif.");
		this.val = val;
	}	
	
	public Nat creerZero() {
		return this.creerZero();
	}
	
	public Nat creerSuccesseur(Nat predecesseur) {
		return this.creerNatAvecValeur(predecesseur.val() + 1);
	}
	
	public Nat creerNatAvecRepresentation(String repDecimal) {
		return this.creerNatAvecValeur(Integer.parseInt(repDecimal));
	}
	
	public int val() {
		return this.val;
	}
	
	public boolean estNul() {
		return this.val() == 0;
	}
	
	public Nat predecesseur() {
		if(this.estNul())
			throw new UnsupportedOperationException("Pas de prédécesseur.");
		return this.creerNatAvecValeur(this.val() - 1);
	}

	public int chiffre(int i) {
		return (i == 0) ? this.val()%10 : (this.creerNatAvecValeur(this.val()/10)).chiffre(i-1);
	}
	
	public int taille() {
		return (this.val() < 10) ? 1 : 1 + (this.creerNatAvecValeur(this.val() / 10)).taille();
	}
}

File: typescript/session2/td/agregation/IntPositif.ts
import { EtatNaturelPur, EtatNaturelPur_defaut } from "./EtatNaturelPur";
import { heriterInterfacesConcretes } from "../../../heritageMultiple/heritageInterfacesconcretes";

export class IntPositif implements EtatNaturelPur {
    constructor(private valeur : number) {
        if (valeur < 0){
            throw new Error("Pas de Nat a partir d'un int negatif");
        }
    }

    val() : number {
        return this.valeur;
    }

    estNul() : boolean {
        return this.val() === 0;
    }

    predecesseur() : EtatNaturelPur {
        if(this.estNul) {
            throw new Error("Pas de prédécesseur");
        }
        return this.creerNatAvecValeur(this.val() - 1);
    }

    chiffre(i : number) : number{
		return (i == 0) ? this.val()%10 : (this.creerNatAvecValeur(this.val()/10)).chiffre(i-1);
    }

    taille() : number {
		return (this.val() < 10) ? 1 : 1 + (this.creerNatAvecValeur(this.val() / 10)).taille();
    }

    creerNatAvecValeur(valeur : number) : EtatNaturelPur {
        return new IntPositif(valeur);
    }

    representation() : string {
        return String(this.val());
    }

    estEgal(x : EtatNaturelPur) : boolean {
        return x.val() === this.val();
    }

    creerZero : () => EtatNaturelPur;
    creerSuccesseur : (predecesseur : EtatNaturelPur) => EtatNaturelPur;
    creerNatAvecRepresentation : (repDecimal : string) => EtatNaturelPur;
}

heriterInterfacesConcretes(IntPositif, [EtatNaturelPur_defaut])